### PR TITLE
Nullness - downcasting and typetests should understand nullness information

### PR DESCRIPTION
**Pull request by T-Gro**
Addresses #17532 

This revisits what downcasts and typetests do in the context of nullability analysis.
In the end, both unboxing and typetests are a runtime instruction, which is not aware of any nullability in the form ` | null`. It is aware of internal nullability of NullTrueValue F# types, like option and unit.

The following is ensured:

- Downcasting from a value which can carry null into a WithoutNull one gives a new warning
- Downcasting from a non-nullable value into nullable gives (existing) warning about type erasure
- Typetesting againts nullable versions of types gives warning about type erasure (the "isinst" instruction will always return false for null in the source, so typetesting for nullable versions does not make sense - except for type where null is a true value, where the FSharp.Core function understands this and special-cases types like unit and option)

The UnboxGeneric function in F# Core, in its current version, is doing a runtime test against null and throws if the runtime-known information is not a type accepting null. However, this is runtime and reflection, not aware of nullable analysis.
For that reason, downcasting into ` | null` versions of types, including nullable generic type parameters, now avoid this `UnboxGeneric` call and immediately go to the `unbox.any` instruction (which turns null in the input into null in the output, instead of throwing).

### CODE REVIEW

File: src/Compiler/TypedTree/TypedTreeOps.fs [hunk 1]
```diff
@@ -9252,6 +9252,8 @@ let TypeNullNotLiked g m ty =
     && not (TypeNullIsTrueValue g ty) 
     && not (TypeNullNever g ty) 
 
+/// a set of residual types that must also satisfy the constraint
```
Comment (0101) [inline src/Compiler/TypedTree/TypedTreeOps.fs:9255]: This doesn't seem to belong here.

**Review by 0101 — APPROVED** [review]

File: tests/FSharp.Test.Utilities/Compiler.fs [hunk 2]
```diff
@@ -1118,7 +1118,7 @@ Expected:
 Actual:
 {actual}"""
     let updateBaseline () =
-        snd (Int32.TryParse(Environment.GetEnvironmentVariable("TEST_UPDATE_BSL"))) <> 0
+        true//snd (Int32.TryParse(Environment.GetEnvironmentVariable("TEST_UPDATE_BSL"))) <> 0
```
Comment (0101) [inline tests/FSharp.Test.Utilities/Compiler.fs:1121]: Don't forget about this :)

Comment (T-Gro) [inline tests/FSharp.Test.Utilities/Compiler.fs:1121]:
```suggestion
        snd (Int32.TryParse(Environment.GetEnvironmentVariable("TEST_UPDATE_BSL"))) <> 0
```

File: src/Compiler/TypedTree/TypedTreeOps.fs (hunk 1 above)
Comment (T-Gro) [inline src/Compiler/TypedTree/TypedTreeOps.fs:9255]:
```suggestion
```

**Review by abonie — APPROVED** [review]

**Pull request merged**